proposals/pytorch-elastic-proposal.md

```diff
@@ -0,0 +1,436 @@
+}
+
+func (e ElasticEnvVarGenerator) generateEnvRDZVConf(elasticPolicy *pytorchv1.ElasticPolicy) *corev1.EnvVar {
+	if elasticPolicy.RDZVConf == nil {
+		return nil
+	}
+	val := ""
+	for _, conf := range elasticPolicy.RDZVConf {
+		val += fmt.Sprintf("%s=%s,", conf.Key, conf.Value)
+	}
+	return &corev1.EnvVar{
+		Name: EnvRDZVConf,
+		// Remove the last comma.
+		Value: val[:len(val)-1],
+	}
+}
+
+func (e ElasticEnvVarGenerator) generateEnvBackend(elasticPolicy *pytorchv1.ElasticPolicy) corev1.EnvVar {
+	if elasticPolicy.RDZVBackend != nil {
+		return corev1.EnvVar{
+			Name:  EnvRDZVBackend,
+			Value: string(*elasticPolicy.RDZVBackend),
+		}
+	}
+	return corev1.EnvVar{
+		Name:  EnvRDZVBackend,
+		Value: string(pytorchv1.BackendC10D),
+	}
+}
+```
+
+### Ports
+
+The worker spec can have the following ports:
+
+- pytorchjob-port: The port for the rdzv port (if needed).
+
+### Resulting Spec
+
+The resulting worker pod looks like this:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: ${pytorchjob.metadata.name}-worker-1
+spec:
+  containers:
+  - command:
+    - python
+    - -m
+    - torch.distributed.run
+    - /workspace/examples/imagenet.py
+    - --arch=resnet18
+    - --epochs=20
+    - --batch-size=32
+    - --workers=0
+    - /workspace/data/tiny-imagenet-200
+    env:
+    - name: LOGLEVEL
+      value: DEBUG
+    - name: PYTHONUNBUFFERED
+      value: "0"
+    - name: PET_RDZV_ENDPOINT
+      value: elastic-example-imagenet-worker-0:29400
+    - name: PET_RDZV_BACKEND
+      value: c10d
+    - name: PET_NODES
+      value: "1:2"
+    - name: PET_MAX_RESTARTS
+      value: "100"
+    image: gaocegege/pytorch-elastic-example-imagenet:1.0.0-sigterm
+    imagePullPolicy: IfNotPresent
+    name: pytorch
+    ports:
+    - containerPort: 29400
+      name: pytorchjob-port
+      protocol: TCP
+
+```
+
+## Autoscaler Integration
+
+Three fields should be added in CustomResourceDefinition:
+
+```yaml
+    scale:
+      specReplicasPath: .spec.pytorchReplicaSpecs.Worker.replicas
+      # Should we have a total replicas?
+      statusReplicasPath: .status.replicaStatuses.Active
+      labelSelectorPath: .status.labelSelector
+```
+
+`LabelSelector` should be introduced into `common.ReplicaStatus`.
+
+```diff
+type ReplicaStatus struct {
++	// LabelSelector is the selector for the replica.
++	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
+
+	// The number of actively running pods.
+	Active int32 `json:"active,omitempty"`
+
+	// The number of pods which reached phase Succeeded.
+	Succeeded int32 `json:"succeeded,omitempty"`
+	// The number of pods which reached phase Failed.
+	Failed int32 `json:"failed,omitempty"`
+}
+```
+
+Then `PyTorchJob` has the `scale` subResource, then it can work with Autoscaler. The only problem is that `PyTorchJob` already has the minReplicas and maxReplicas fields. They are used to generate commands. The Autoscaler resource needs them, too. Thus users may need to define them again.
```

Comment:
Would it make sense for the pytorch controller to create a HPA object in behalf of the user?

Reply:
Sounds reasonable. Should we do it in PyTorchJob, or should we make a new abstraction, maybe ElasticJob, to manage the underlying PyTorchJob and the HPA?

/cc @kubeflow/wg-training-leads

Reply:
Maybe better start with just PyTorch and add the abstraction later if it ends up working well.